### PR TITLE
Added support for webdataset

### PR DESCRIPTION
**Pull request by robvanvolt**

### CODE REVIEW

**Comment by robvanvolt** [comment]
I was finally able to get WebDataset implemented to DALLE-pytorch - it is completely optional and it won't break any other current functions despite being "early beta" so to say.

Advantages:
* being able to stream data without having to download the whole dataset to your computer
* one line of WebDataset replaces the complete loader.py in the subfolder dalle_pytorch
* distributed training for horovod / deepspeed is also possible, but not yet implemented, and will reduce a good amount of code, as it supports MultiDataset, also a oneliner

If you want to train with WebDataset format, just add the tag --wds to train_dalle.py and add, comma separated, the image column key and the text column key, e.g. --weds img,cap

**Comment by afiaka87** [comment]
@robvanvolt Is this what bliss feels like? Here in a few days when my desktop isn't all tied up I will 100% be converting all of my available datasets to the Webdataset format and re-hosting them. 

Thanks a ton for keeping up with this! Also - thanks for the creation script. I think that and a few other things about WebDataset deserve a place in the wiki here perhaps. I'm taking on a few too many things currently - but I'll try to add things there when I am able to test this.

@lucidrains are you familiar with WebDataset? It's a great project from pytorch but is pre-rfc I believe. I would love to see this merged!

File: train_dalle.py [hunk 1]
```diff
@@ -14,6 +14,11 @@
 from dalle_pytorch.loader import TextImageDataset
 from dalle_pytorch.tokenizer import tokenizer, HugTokenizer, ChineseTokenizer, YttmTokenizer
 
+import io
+import PIL
+import webdataset as wds
```
Comment (chesse20) [inline train_dalle.py:19]: awesome imports great job

**Review by chesse20 — APPROVED** [review]
great pr

**Comment by rom1504** [comment]
@robvanvolt do you think we should wait for the finishing touches on this or merge a first version even if not everything is as good as you want it to be?
How well does it work?

I think a minimal change you can do here is update the README with instructions on how to use this / how to make webdataset format datasets. The current readme is pretty complete and informative, let's keep it that way.

**Comment by robvanvolt** [comment]
> @robvanvolt do you think we should wait for the finishing touches on this or merge a first version even if not everything is as good as you want it to be?
> How well does it work?
> 
> I think a minimal change you can do here is update the README with instructions on how to use this / how to make webdataset format datasets. The current readme is pretty complete and informative, let's keep it that way.

Right, I made the first pull request for people eagerly to test webdataset / try it out without too much effort. The merge to main should not be done before the todos up there are implemented and work flawlessly (probably in 1-2 weeks :))

**Comment by afiaka87** [comment]
@mehdidc when you have an opportunity - would you be able to test this on the copy of conceptual captions 12m (stored in the webdataset format right?) on julich?

**Comment by mehdidc** [comment]
@afiaka87 Definitely, that's on the next steps, it's compatible with WebDataset format but the captions are missing from the tars. We could either just link filenames with captions through a pandas dataframe, or create a new set of tars with captions inside as a file like what they seem to do here: https://webdataset.github.io/webdataset/creating/.

File: train_dalle.py (hunk 1 above)
Comment (robvanvolt) [reply train_dalle.py:19]: Adjusted the imports so they have only the necessary functions imported, thank you for your feedback!:)

**Comment by robvanvolt** [comment]
> @afiaka87 Definitely, that's on the next steps, it's compatible with WebDataset format but the captions are missing from the tars. I think would need a new set of tars where image and captions are provided togeteher, I think adapting @robvanvolt script https://github.com/robvanvolt/DALLE-datasets/blob/e62d2c74107c0ce6b77c84078b8f34a2cb2c5145/wds_create.py would work fine.

I think I found the reason for deepspeed and WebDataset incompatibility... I will run a few tests and commit the changes soon! If it works, I will tidy up the code and test sharding!:)

**Comment by afiaka87** [comment]
> > @afiaka87 Definitely, that's on the next steps, it's compatible with WebDataset format but the captions are missing from the tars. I think would need a new set of tars where image and captions are provided togeteher, I think adapting @robvanvolt script https://github.com/robvanvolt/DALLE-datasets/blob/e62d2c74107c0ce6b77c84078b8f34a2cb2c5145/wds_create.py would work fine.
> 
> I think I found the reason for the incompatibility between deepspeed and WebDataset... and a solution for this... I will run a few tests and if it works, commit the changes soon, tidy up the code and test sharding afterwards!:)
> 
> **Update:** released working deepspeed branch

What was the solution @robvanvolt? Curious

**Comment by robvanvolt** [comment]
> 
> 
> > > @afiaka87 Definitely, that's on the next steps, it's compatible with WebDataset format but the captions are missing from the tars. I think would need a new set of tars where image and captions are provided togeteher, I think adapting @robvanvolt script https://github.com/robvanvolt/DALLE-datasets/blob/e62d2c74107c0ce6b77c84078b8f34a2cb2c5145/wds_create.py would work fine.
> > 
> > 
> > I think I found the reason for the incompatibility between deepspeed and WebDataset... and a solution for this... I will run a few tests and if it works, commit the changes soon, tidy up the code and test sharding afterwards!:)
> > **Update:** released working deepspeed branch
> 
> What was the solution @robvanvolt? Curious

Basically I just added a custom dataloader instead of the distr_dl provided by deepspeed, I went over an issue in the DeepSpeed where they state it is not necessary to use the deepspeed dataloader!:)

**Comment by robvanvolt** [comment]
Streaming as well was loading multiple tar files seem to work flawlessly now!:)

**Comment by rom1504** [comment]
Streaming looks super cool

**Comment by afiaka87** [comment]
Fantastic work as usual @robvanvolt - the streaming is a game changer. I need to know how well that works and whether or not I could just offload my datasets entirely to my amsterdam server and free up space locally. That would be fanatastic.

**Comment by afiaka87** [comment]
Let's make sure this runs on this notebook before merging as well.

https://colab.research.google.com/gist/afiaka87/bbac0038b213a82c067b4766cdf45e0d/ci_train_dalle_pytorch_wip.ipynb

**Comment by robvanvolt** [comment]
> Let's make sure this runs on this notebook before merging as well.
> 
> https://colab.research.google.com/gist/afiaka87/bbac0038b213a82c067b4766cdf45e0d/ci_train_dalle_pytorch_wip.ipynb

Works fine with your colab! :-)

https://colab.research.google.com/drive/15lRQl_NlzCwaEdnGE-jcYM23G7rqCqO5?usp=sharing

Note that the Nvidia Dataset given as an example does provide json format captions and not string captions, so training does not make sense there (even though the loss decreases, as the whole json file will be seen as a string caption).

Also Apex runs fine!

Note that in theory, the webdataset implementation should not break anything, as all the changes that were made take only effect when "ENABLE_WEBDATASET" is set to true - without this tag, everything runs as expected from the main branch.

**Comment by afiaka87** [comment]
@robvanvolt I think we should set this up as a default if we do a redesign btw. Need any help with the merge?

**Comment by robvanvolt** [comment]
> 
> 
> @robvanvolt I think we should set this up as a default if we do a redesign btw. Need any help with the merge?

I just looked over it - no problem merging it to main! Will do it today in the evening.:)

File: train_dalle.py [hunk 2]
```diff
@@ -82,13 +106,13 @@
 
 model_group.add_argument('--dim', default = 512, type = int, help = 'Model dimension')
 
-model_group.add_argument('--text_seq_len', default = 256, type = int, help = 'Text sequence length')
+model_group.add_argument('--text_seq_len', default = 128, type = int, help = 'Text sequence length')
```
Comment (rom1504) [inline train_dalle.py:109]: Looks like you changed the defaults @robvanvolt 
Did you mean to ?
Otherwise let's revert this part

**Comment by rom1504** [comment]
Just a small note on GitHub-fu that I think is relevant @lucidrains : I think it's much nicer to used the "squash and merge" option of GitHub rather than the "merge pr" option when merging.
It results in a single commit in master history and the author is still the PR commit authors.

It then makes it much clearer to look at master history and understand what happened since it's all linear and simple.

File: train_dalle.py (hunk 2 above)
Comment (rom1504) [reply train_dalle.py:109]: #311

Comment (robvanvolt) [reply train_dalle.py:109]: I did it for training on my older computers, as I always ran out of RAM, so it was much easier for me with the default being 128 - I don't mind, we can revert it again!:)
